### DISCUSSION TITLE
Inconsistent behavior of marking a value as constant and type alias in pyright and pylance

### DISCUSSION THREAD

**Discussion by notcome** [microsoft/pyright#7242] [Q&A] [answered]
I have a code where I generate a dynamic Pydantic BaseModel. I would have a constant that contains a member of the type `type[BaseModel]`. And here I noticed some pretty interesting behaviors. Consider the following code:

```python
from pydantic import BaseModel


class Wrapper:
    model: type[BaseModel]


def make_base_model() -> type[BaseModel]:
    raise NotImplementedError()


T1 = make_base_model()


def make_base_model_wrapped() -> Wrapper:
    raise NotImplementedError()


T2 = make_base_model_wrapped().model


def call1(t: T1):
    print(t.model_dump_json())
    print(t.random())
    return


def call2(t: T2):
    print(t.model_dump_json())
    print(t.random())
    return
```

In VS Code with Pylance, when I hover T1/T2, I noticed that:

- T1 is classified as a constant
- T2 is classified as a type alias

Pylance would complain about the signature for `call1`, for variable not allowed in type expression. But it happily type checks `call2`. In both cases though, Pyright complains.

I have a few questions.

- Why is such an inconsistency between Pyright and Pylance?
- Why does Pylance perform differently for `call1` and `call2`? Is this a bug?
- Is there a way to get some type checking in my case? For `call2/T2`, since Pylance thinks `T2` is a type alias, it actually performs type checking and hence it warns about the non-existent `random` method call. But I cannot get this in Pyright, or for `call1/T1` in VS Code.

In my particular use case, I am using this for FastAPI. I would like to have documentation generated for my endpoint, so using a value as type signature is actually needed. I could create another function that takes a `BaseModel` to implement the actual endpoint to make things work for Pyright, but that feels a little bit clumsy…

**Comment by erictraut** [comment, marked as answer]
The inconsistency is because pylance is using a version of pyright that's a few versions old. If you upgrade to the latest prerelease version of Pylance (or if you wait a few weeks until the next version of Pylance is published), you should see the same behavior.

The typing community has been working to clear up ambiguities on the [Python typing spec](https://typing.readthedocs.io/en/latest/spec/index.html), and I've been working to tighten up [pyright's conformance](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) with this spec. 

The typing spec doesn't allow for the use of variables within type annotation expressions, like you're doing in the code above. It's fine to use `BaseModel` in a type expression, so that's what I'd suggest in your case. Alternatively, you could define a protocol class (a structural type) that defines the core methods and attributes of the type you are dynamically creating.

Depending on how you intend to use the `Wrapper` class, you might want to consider making it generic, like this:

```python
class Wrapper[T: BaseModel]:
    model: type[T]
```

**Reply by notcome** [reply]
Thanks! I am using Cursor and somehow I can't get to the latest Pylance version. That must be it.

In my case, my model class is auto generated via a decorator from an arbitrary function's signature, so I don't really see a way to bypass it. I guess I will settle with a `type: ignore` and an inner function taking a plain `BaseModel`.
